free.cabal

```diff
@@ -81,25 +81,40 @@ library
     GADTs
 
   build-depends:
-    base                 == 4.*,
-    comonad              >= 4 && < 6,
-    containers           < 0.7,
-    distributive         >= 0.2.1,
-    exceptions           >= 0.6 && < 0.11,
-    indexed-traversable  >= 0.1.1 && < 0.2,
-    mtl                  >= 2.0.1.0 && < 2.3,
-    profunctors          >= 4 && < 6,
-    semigroupoids        >= 4 && < 6,
-    th-abstraction       >= 0.4 && < 0.5,
-    transformers         >= 0.2.0   && < 0.6,
-    transformers-base    >= 0.4 && < 0.5,
+    base                 >= 4.5     && < 5,
+    comonad              >= 5.0.8   && < 6,
+    containers           >= 0.3     && < 0.7,
+    distributive         >= 0.5.2   && < 1,
+    exceptions           >= 0.10.4  && < 0.11,
+    indexed-traversable  >= 0.1.1   && < 0.2,
+    semigroupoids        >= 5.3.5   && < 6,
+    th-abstraction       >= 0.4.2.0 && < 0.5,
+    transformers         >= 0.3     && < 0.6,
+    transformers-base    >= 0.4.5.2 && < 0.5,
     template-haskell     >= 2.7.0.0 && < 2.18
 
+  -- GHC-7.8 bundles transformers-0.3,
+  -- mtl-2.2.* requires transformers >=0.4
+  if impl(ghc >=7.10)
```

Comment:
What is the benefit of having a conditional dependency where both branches of the `if` declare a dependency on the same package? I would hope that `cabal`'s constraint solver would be smart enough to pick the appropriate version of `mtl` depending on which GHC is being used. But perhaps there's something deeper here that I'm missing.

Reply:
It's more restricted for older GHCs. Specifically, `mtl-2.2.1` isn't *explicitly safe*, so causes warnings. The dep definitions here are a bit complicated, but excludes (nasty, no-one should need) corner-cases.